### PR TITLE
Refactored the classical tiling calculator

### PR DESCRIPTION
**Pull request by micheles**
This is more efficient than before, it logs correctly on the controller node and it does not instantiate GSIMs on the worker nodes. This requires https://github.com/gem/oq-hazardlib/pull/406. The tests are green: https://ci.openquake.org/job/zdevel_oq-risklib/1269

### CODE REVIEW

File: openquake/calculators/classical.py [hunk 1]
```diff
@@ -151,43 +155,43 @@ def classical(sources, sitecol, rlzs_assoc, monitor):
     truncation_level = monitor.oqparam.truncation_level
     imtls = monitor.oqparam.imtls
     trt_model_id = sources[0].trt_model_id
+    # sanity check: the trt_model must be the same for all sources
+    for src in sources[1:]:
+        src.trt_model_id == trt_model_id
```
Comment (ptormene) [inline openquake/calculators/classical.py]: Please add `assert`

File: openquake/calculators/classical.py [hunk 2]
```diff
@@ -151,43 +155,43 @@ def classical(sources, sitecol, rlzs_assoc, monitor):
...
-    for key in val:
-        if key == 'calc_times':
-            acc[key].extend(val[key])
-        elif key == 'bbs':
-            for bb in val['bbs']:
-                acc['bb_dict'][bb.lt_model_id, bb.site_id].update_bb(bb)
-        else:  # aggregate curves
-            acc[key] = agg_curves(acc[key], val[key])
-    return acc
+    if len(array) == n:  # nothing to expand
```
Comment (ptormene) [inline openquake/calculators/classical.py]: If `len(array) > n`, I guess we should raise an error

File: openquake/calculators/classical.py [hunk 3]
```diff
@@ -151,43 +155,43 @@ def classical(sources, sitecol, rlzs_assoc, monitor):
...
     for gsim, curves in zip(gsims, curves_by_gsim):
         dic[trt_model_id, str(gsim)] = curves
     return dic
 
 
-def agg_dicts(acc, val):
+def expand(array, n, aslice):
     """
-    Aggregate dictionaries of hazard curves by updating the accumulator
+    Expand a short array to size n by adding zeros outside of the slice
```
Comment (ptormene) [inline openquake/calculators/classical.py]: A simple example would make this more intuitive

File: openquake/calculators/classical.py [hunk 4]
```diff
@@ -222,6 +226,24 @@ class ClassicalCalculator(base.HazardCalculator):
     core_func = classical
     source_info = datastore.persistent_attribute('source_info')
 
+    def agg_dicts(self, acc, val):
+        """
+        Aggregate dictionaries of hazard curves by updating the accumulator
+        """
```
Comment (ptormene) [inline openquake/calculators/classical.py]: Please describe parameters

File: openquake/calculators/classical.py [hunk 5]
```diff
@@ -365,61 +381,22 @@ def _store(self, name, curves, rlz, **kw):
...
     """
-    :param acc: AccumDict (trt_id, gsim) -> N curves
-    :param curves_by_trt_gsim: AccumDict (trt_id, gsim) -> T curves
+    Returns True on tectonic region types
+    which ID in contained in the result_dict.
 
-    where N is the total number of sites and T the number of sites
-    in the current tile. Works by side effect, by updating the accumulator.
+    :param result_dict: a dictionary with keys (trt_id, gsim)
     """
```
Comment (ptormene) [inline openquake/calculators/classical.py]: Please add `:param trt_model:`

**Comment by ptormene** [comment]
Apart from a missing check and some minor notes about docstrings, LGTM

File: openquake/calculators/classical.py (hunk 2 above)
Comment (micheles) [inline openquake/calculators/classical.py]: Done

File: openquake/calculators/classical.py (hunk 3 above)
Comment (micheles) [inline openquake/calculators/classical.py]: Added example

File: openquake/calculators/classical.py (hunk 4 above)
Comment (micheles) [inline openquake/calculators/classical.py]: Done

File: openquake/calculators/classical.py (hunk 5 above)
Comment (micheles) [inline openquake/calculators/classical.py]: Added

**Pull request merged**